File: java/src/main/behavioral/mediator/ConcreteMediator.java
package behavioral.mediator;

import java.util.ArrayList;
import java.util.List;

// ConcreteMediator: Manages communication between colleagues
public class ConcreteMediator implements Mediator {
    private final List<Colleague> colleagues = new ArrayList<>();

    public void registerColleague(Colleague colleague) {
        colleagues.add(colleague);
    }

    @Override
    public void sendMessage(String message, Colleague sender) {
        for (Colleague colleague : colleagues) {
            if (colleague != sender) {
                colleague.receiveMessage(message);
            }
        }
    }
}

// Example Usage:
// public class Main {
//     public static void main(String[] args) {
//         ConcreteMediator mediator = new ConcreteMediator();
//
//         ConcreteColleagueA colleagueA = new ConcreteColleagueA(mediator);
//         ConcreteColleagueB colleagueB = new ConcreteColleagueB(mediator);
//
//         mediator.registerColleague(colleagueA);
//         mediator.registerColleague(colleagueB);
//
//         colleagueA.send("Hello from A!");
//         colleagueB.send("Hi A, this is B!");
//
//         // Expected Output:
//         // ConcreteColleagueA: Sending message -> Hello from A!
//         // ConcreteColleagueB received message: Hello from A!
//         // ConcreteColleagueB: Sending message -> Hi A, this is B!
//         // ConcreteColleagueA received message: Hi A, this is B!
//     }
// }
//
// Common Mistakes to Avoid:
//
// 1. Overcomplicating the Mediator Logic:
//    - Mistake: Adding too much logic inside the mediator.
//    - Fix: Keep the mediator focused on communication, not business logic.
//
// 2. Tight Coupling Between Mediator and Colleagues:
//    - Mistake: The mediator and colleagues being too dependent on each other.
//    - Fix: Use interfaces to decouple implementations.
//
// 3. Not Managing Colleague Registration Properly:
//    - Mistake: Forgetting to register colleagues with the mediator.
//    - Fix: Always register colleagues before they send messages.
//
// 4. Inefficient Broadcasting:
//    - Mistake: Sending messages to all colleagues unnecessarily.
//    - Fix: Filter recipients properly to avoid unnecessary message processing.

File: python/src/behavioral/mediator/concrete_mediator.py
from .mediator import Mediator
from .concrete_colleague_a import ConcreteColleagueA
from .concrete_colleague_b import ConcreteColleagueB


class ConcreteMediator(Mediator):
    """
    Concrete Mediator coordinates communication between colleagues.
    """

    def __init__(self,
                 colleague_a: ConcreteColleagueA,
                 colleague_b: ConcreteColleagueB):
        self.colleague_a = colleague_a
        self.colleague_b = colleague_b

        # Set mediator for colleagues
        self.colleague_a.set_mediator(self)
        self.colleague_b.set_mediator(self)

    def notify(self, sender: object, event: str) -> str:
        if event == "A_ACTION":
            return "Mediator reacts to A and triggers B." + \
                   " | " + self.colleague_b.do_action_b_handler()
        elif event == "B_ACTION":
            return "Mediator reacts to B and triggers A." + \
                   " | " + self.colleague_a.do_action_a_handler()
        return "Unknown event."


# Extend colleagues with handlers for mediator-triggered actions

def _colleague_a_handler(self) -> str:
    return "Colleague A handling mediator request."


def _colleague_b_handler(self) -> str:
    return "Colleague B handling mediator request."


ConcreteColleagueA.do_action_a_handler = _colleague_a_handler
ConcreteColleagueB.do_action_b_handler = _colleague_b_handler


# Example Usage:
#
# from concrete_colleague_a import ConcreteColleagueA
# from concrete_colleague_b import ConcreteColleagueB
# from concrete_mediator import ConcreteMediator
#
# a = ConcreteColleagueA()
# b = ConcreteColleagueB()
# mediator = ConcreteMediator(a, b)
#
# print(a.do_action_a())
# print(b.do_action_b())
#
# Output:
# Mediator reacts to A and triggers B. | Colleague B handling mediator request.
# Mediator reacts to B and triggers A. | Colleague A handling mediator request.
#
#
# Common Mistakes to Avoid:
#
# 1. Putting too much logic in colleagues instead of the mediator.
#    - Fix: The mediator should coordinate interactions.
#
# 2. Creating a God Object mediator.
#    - Fix: Keep it focused on communication, not business logic.
#
# 3. Directly coupling colleagues to each other.
#    - Fix: All communication should go through the mediator.
